Feat rt validation upload

machow opened the pull request:
This PR adds convenient functions for fetching our RT data, and a wrapper to run the https://github.com/CUTR-at-USF/gtfs-realtime-validator on it.

* See README on this PR branch: https://github.com/cal-itp/gtfs-rt-validator-api/tree/feat-rt-validation-upload
* Runs simple integration tests of validating data in a test bucket
* Note that this library needs to do things like suffixing a timestamp at the end of filenames, in order for the validator to correctly order them.
* On github release, an action builds and publishes a docker image to the github container registry: https://github.com/orgs/cal-itp/packages?repo_name=gtfs-rt-validator-api

evansiroky commented on gtfs_rt_validator_api.py:
```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
+        
+    """
 
     import gcsfs
```
Seems like if this is imported in this file, that it should exist as a entry within `setup.py`'s `install_requires`.

machow replied:
added

evansiroky commented on README.md:
````diff
@@ -1,19 +1,125 @@
+### Validating a folder
+
+```python
+```
````
Can this be filled out, or the section removed?

evansiroky commented on gtfs_rt_validator_api.py:
```diff
@@ -4,19 +4,66 @@
+    raise NotImplementedError()
+
+
 def validate(gtfs_file, rt_path, out_file=None, verbose=False):
```
Is the `out_file` argument needed?

evansiroky commented on gtfs_rt_validator_api.py:
```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
+
+        dst_parent.mkdir(parents=True, exist_ok=True)
+
+        # TODO: check that this works
```
Does it work?

machow replied:
whoops--yes, It's run in the integration tests!

evansiroky commented on gtfs_rt_validator_api.py:
```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
+        gtfs_rt_glob_path: path that GCSFileSystem.glob can uses to list all RT files.
+            Note that this is assumed to have the form {datetime}/{itp_id}/{url_number}/filename.
+        out_dir: a directory to store fetched files and results in.
+        results_dir: a bucket path to copy results to.
```
Doc param name doesn't match function name, see recommended change:

```suggestion
        results_bucket: a bucket path to copy results to.
```

machow replied:
should be updated

evansiroky commented on gtfs_rt_validator_api.py:
```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
-        # TODO: replace with gtfs_rt_glob
-        if gtfs_rt_date:
-            fs.glob(f"{gtfs_rt_date}T*/")
+        # TODO: remove hard-coding
```
Is this comment still needed?

machow replied:
removed!

evansiroky commented on gtfs_rt_validator_api.py:
```diff
@@ -25,42 +72,169 @@ def validate(gtfs_file, rt_path, out_file=None, verbose=False):
+
+            fs.put(final_files, results_bucket)
+
+    except Exception as e:
+        if isinstance(tmp_dir, TemporaryDirectory):
+            tmp_dir.cleanup()
+
+        raise e
```
Is the temporary directory cleanup still needed even if an exception occurs? If so, that cleanup should happen in a finally clause:

```suggestion
    except Exception as e:
        raise e
    finally:
        if isinstance(tmp_dir, TemporaryDirectory):
            tmp_dir.cleanup()
```

machow replied:
Ah, thanks

evansiroky commented on tests/test_download.py:
```diff
@@ -0,0 +1,39 @@
+        )
+
+        print("validating")
+        validate(zip_schedule, dir_rt)
```
This test doesn't contain any assertions to determine whether the resulting output is what was expected. Please add at least one assertion per test.

machow replied:
refactored tests

* added script for seeding data to gcs bucket
* now create a temporary cloud bucket per test as needed
* added assertions

evansiroky commented on tests/test_download.py:
```diff
@@ -0,0 +1,39 @@
+        "gs://calitp-py-ci/gtfs-rt-validator-api/gtfs_rt_126/2021*/126/0/*",
+        out_dir="tests/data/validate_gcs_bucket",
+        results_bucket="gs://calitp-py-ci/gtfs-rt-validator-api/test_output"
+    )
```
This test doesn't contain any assertions to determine whether the resulting output is what was expected. Please add at least one assertion per test.

machow replied:
refactored tests to add assertion checking for results.json files

evansiroky reviewed (changes requested):
See various comments, but it's looking like it ought to work well.

machow commented:
Thanks for the feedback--I think something got screwed up with the tests. I refactored to make them work even for concurrent runs of the test suite, and added assertions + the data seeding script :)

evansiroky commented on script/seed_bucket_data.py:
```diff
@@ -0,0 +1,18 @@
+from calitp.storage import get_fs
```
Please write some comments at the top of this file that explain the purpose of this file and how to use it. It might be a good idea to add a blurb in the README about this too.

machow replied:
done!

evansiroky commented on tests/test_download.py:
```diff
@@ -0,0 +1,73 @@
+from calitp.storage import get_fs
+from pathlib import Path
+
+GCS_BASE_DIR="gs://calitp-py-ci/gtfs-rt-validator-api"
```
It would be helpful to add a comment noting the existence of the `seed_bucket_data.py` script that can help add data testing data to this directory.

machow replied:
done, thanks!

evansiroky reviewed (changes requested):
That seeding script ought to be helpful. Please add some comments/docs about that script and then this should be good to go.
